### PR TITLE
Discussion on NNCache and TTable

### PR DESCRIPTION
**Pull request by killerducky**
I made a pull so there is code to see for example, but first I want to check what people think of this. This isn't something ready to be merged as-is. There are three commits with two major changes in this pull that are IMO related. The first commit just adds some debug prints. Second commit, I changed NNCache to use board.get_hash instead of the feature planes. This allows NNCache to treat positions that should in theory have the same value, even if the previous 8 moves are not identical, as identical. Note Simple-Ko differences are covered by board.get_hash. Super-Ko is handled by the tree search. I believe I saw some discussion where gcp was in favor of this.

Stats of baseline:
`NNCache: 400/356394 hits/lookups = 0.1% hitrate, 355994 inserts, 6000 size, 0 collisions
`
Stats with NNCache change:
`NNCache: 40400/451679 hits/lookups = 8.9% hitrate, 411279 inserts, 6000 size, 0 collisions
`

Third commit I disabled the TTable by just commenting out the calls to it. I don't think the TTable is doing what it really should. It does a good job of aggregating m_visits and m_eval_sum to get a more accurate eval for the equivalent nodes. But that information is not propagated up to the parents. A second problem is since the m_visits of the equivalent nodes rises faster than their sibilings, the UCTSearch will be biased *against* exploring nodes that are in the TTable. And those extra visits are not being propagated up to the respective parents. I can't think of an easy way to fix this, and it seems by using the above NNCache change we can simply have the CPU eventually re-simulate everything for very low cost by hitting the NNCache as it moves down the tree from equivalent nodes.

Some evidence of the bias against exploring nodes that are in the TTable:
Stats with NNCache change and TTable **on** (first move only):
```
NNCache: 100/178 hits/lookups = 55.9% hitrate, 78 inserts, 78 size, 0 collisions
NNCache: 200/367 hits/lookups = 54.3% hitrate, 167 inserts, 167 size, 0 collisions
NNCache: 300/566 hits/lookups = 52.9% hitrate, 266 inserts, 266 size, 0 collisions
NNCache: 400/801 hits/lookups = 49.9% hitrate, 401 inserts, 401 size, 0 collisions
NNCache: 500/1019 hits/lookups = 49.0% hitrate, 519 inserts, 519 size, 0 collisions
NNCache: 600/1211 hits/lookups = 49.5% hitrate, 611 inserts, 611 size, 0 collisions
NNCache: 700/1423 hits/lookups = 49.2% hitrate, 723 inserts, 723 size, 0 collisions
```
Stats with NNCache change and TTable **off** (first move only):
```
NNCache: 100/176 hits/lookups = 56.5% hitrate, 76 inserts, 76 size, 0 collisions
NNCache: 200/342 hits/lookups = 58.3% hitrate, 142 inserts, 142 size, 0 collisions
NNCache: 300/503 hits/lookups = 59.5% hitrate, 203 inserts, 203 size, 0 collisions
NNCache: 400/667 hits/lookups = 59.9% hitrate, 267 inserts, 267 size, 0 collisions
NNCache: 500/830 hits/lookups = 60.2% hitrate, 330 inserts, 330 size, 0 collisions
NNCache: 600/1020 hits/lookups = 58.8% hitrate, 420 inserts, 420 size, 0 collisions
NNCache: 700/1200 hits/lookups = 58.3% hitrate, 500 inserts, 500 size, 0 collisions
NNCache: 800/1394 hits/lookups = 57.3% hitrate, 594 inserts, 594 size, 0 collisions
NNCache: 900/1565 hits/lookups = 57.5% hitrate, 665 inserts, 665 size, 0 collisions
```
You can see the TTable on case there are less NNCache hits. I believe it's because of the bias against exploring nodes in the TTable.

Looking for feedback on if we want to try something like this, or if you would like to see different experiments such as head to head tournament etc.

### CODE REVIEW

File: src/NNCache.cpp [hunk 1]
```diff
@@ -55,28 +39,26 @@ bool NNCache::lookup(const Network::NNPlanes& features, Network::Netresult & res
     }
 
     const auto& entry = iter->second;
-    if (entry->features != features) {
-        // Got a hash collision.
-        ++m_collisions;
-        return false;
-    }
+    // TODO: I don't think we can check for collisions with this method
```
[the remark this answers is not in the document]
Comment (killerducky) [reply src/NNCache.cpp]: Hehe. Well I realized right after pushing this that it's easy enough to check that the stones on the board are the same. Not sure how careful we should be about this.

**Comment by roy7** [comment]
Random thoughts. This is above my pay grade for now since I've never really dived in LZ's code.

One of issues about data not flowing up to the parents, let me know if I have this correct. Can something like this happen. Some prior move has a bunch of search in a specific board state. Say A B C D gets us there and all of our search is in that path. However the opponent went A D. We search and decide C E is the most likely result and spend all of our time there, never re-discovering all of the data we have at A D C B (same board state as A B C D for this example).

So all of the prior search at A B C D which would be available at A D if we searched C B is forgotten. This is the sort of problem you originally posted about in the other thread.

Additionally, we have a different issue if I understand properly. Say we spend tons of time searching A B C D. Call it 100K visits. Then we decide to start searching A D X X instead. Eventually we explore into A D C B and have a TTable hit. This adds 100K visits to this path since we're adding in the TTable data. A D C B now has a lot more than before, making it the most likely move to take and as we explore more into A D C B the A B C D value doesn't change, even though the TTable for position A B C D (same as A D C B) is gaining more info as we search here. But this new info will never populate back to node A B C D in the tree (in fact, will a node ever lookup it's TTable additional times after the first fetch in case some other path updates that TTable's data?). Or does the tree search actually merge A B C D and A D C B into the same node internally? I assume not since you discuss NNCache being able to view these as the same with different prior histories leading to them.

Where this leaves me concerned is what we saw in your other thread. A low prior move was found to be a very strong candidate move, but in a later search it was never expanded once (was it biased against as already being in the TTable?) and thus that information was lost. Adding insult to injury, the 2nd best move shows as a 0 target in dump_training as well, making it harder to relocate in the future when the prior should have been increasing instead. 

Over a long period of time and millions of games, this could have an effect where for a given board state, low prior but critical moves (hard to see on heatmap but important to learn) are forced further and further towards 0 when dump_training doesn't see the TTable hits. No matter how high or important that node turned out to be, if the prior is low enough all of the TTable data from prior search turns might be lost on the following turns. Each time the prior is trained lower, it'll make it harder to discover in future games as well. Eventually making it impossible for LZ to see certain low prior but critical moves since they've been trained all the way down to 0? 

To address the forward looking problem, @gcp did mention elsewhere maybe all children should have lookup done in the TTable each time a new node is expanded. But with the new NNCache and tree re-use, it might be cleaner to just stop using TTable?

I didn't read the tree-reuse PR and I don't know how that works, or how tree searching in general is currently done. But I wonder if we're basing everything off NNCache, and if NNCache is very fast to check for hits, should we (if we don't already) check every child of a newly expand mode for NNCache hits? If it's a miss, do nothing. If it's a hit, add it to the tree and recursively expand that node. This way when we start search 'for real' we have a full expanded search tree to begin with, and any past data under our current search paths has been inherited. Of course, the real tree re-use already in place is probably much more elegant than this and just inherits branch as the new root with no re-expansion. :)

**Comment by roy7** [comment]
I think if nothing else, NNCache being based on position and simple ko only and thus sharing data with other identical board states with different histories, is a nice improvement.

**Comment by billyswong** [comment]
Although GCP has clarified that "reduced number of bit planes (from inputting [t0, t-7] to [t0, t-2])" does not help on computational time in https://github.com/gcp/leela-zero/issues/591#issuecomment-358068524 , I still think it is an ultimately cleaner implementation than all these ko-hashes and node data merging/sharing, as judging which and how nodes can be merged or shared involves specific human game knowledge.

**Comment by roy7** [comment]
BTW,

> NNCache: 10000/112337 hits/lookups = 8.9% hitrate, 102310 inserts, 6000 size, 0 collisions

When I get a status like that, since size is only 6000 does that mean only 6K of the 102K nodes are stored?

**Comment by gcp** [comment]
>Second commit, I changed NNCache to use board.get_hash instead of the feature planes. This allows NNCache to treat positions that should in theory have the same value, even if the previous 8 moves are not identical, as identical. Note Simple-Ko differences are covered by board.get_hash. Super-Ko is handled by the tree search. I believe I saw some discussion where gcp was in favor of this.

If you think I was in favor of this there must've been some misunderstanding.

What you say is wrong. That is, moves with the same hash are most definitely NOT guaranteed to be positions that should in theory have the same value. We fixed the most common case of a simple ko but all kinds of superkos will still go wrong. The NNCache currently does guarantee (aside from hash collisions) that the results will be the same, because it considers all the input planes that the network would.

The search just stops you from exploring moves that are illegal due to superko, it won't do anything to stop NNCache now from giving "wrong" results.

Now, just because the reasoning is wrong, does not mean the proposed change is. But "reasoning" about this is pointless. You need to run a validation test and see if it becomes stronger (or no weaker but faster).

If it's stronger, who cares if its wrong!

>the UCTSearch will be biased against exploring nodes that are in the TTable

What else would the point of the TTable be?! It's to detect transpositions or positions we already visited before, copy in the existing result, and search *elsewhere*.

**Comment by killerducky** [comment]
Ok I'm running some tests now.

I agree the TTable should copy the results and search elsewhere. I just think the new information is not shared with the parent and upwards the way it should be. I have an idea how to share the information in a better way, but it requires adding a couple extra members to NNode. Are you going to take #574?

**Comment by gcp** [comment]
#574 is incomplete so I can't take it for now, but the general idea looks good.

**Comment by killerducky** [comment]
I'm brainstorming ideas, one idea I had involves having TTable store m_canonical_node which holds the head of the subtree from that point down. This would allow the entire subtree to be replicated when there is a TTable hit. It would do copies and each subtree has their own m_children to maintain clear ownership. 

The only ownership issue is TTable.m_canonical_node. That could be handled by making TTable.m_canonical_node a raw pointer. Change TTable from global to owned by UCTSearch to support multiple UCTSearches. Modify set_gamestate to also invalidate the TTable, because that is the moment when canonical_node can become invalid.

Another way is using shared_ptrs for all the nodes, but maybe that is too expensive?

Of course I would test if this produces results but does it seem feasible to you?

**Comment by gcp** [comment]
> That could be handled by making TTable.m_canonical_node a raw pointer. 

This is always the wrong solution. Sounds like an absolute nightmare in lifetime management.

>Change TTable from global to owned by UCTSearch to support multiple UCTSearches. 

I don't see what this would gain.

>Another way is using shared_ptrs for all the nodes, but maybe that is too expensive?

I don't think so. Specifically:

>It would do copies and each subtree has their own m_children to maintain clear ownership.

Copying shared_ptrs would be a lot cheaper than copying subtrees!

**Comment by killerducky** [comment]
I have some preliminary results, both versions are slightly stronger than base, but not statistically significant yet. Also I ran these using `-t 4` and set them to think 0.5s (by using `-l 500` and giving it 1s per move). I picked 0.5s so lzbase averages 1600 playouts. Both changes cause it to run faster, but without using the extra time to think more it is weaker (I corrupted my ringmaster results showing this, I'll try to reconstruct it). 

If something like this were added I guess it should be with a command line switch, otherwise the current run will suffer. Even assuming the current 53.7% winrate holds up as I get more results, I don't know if that is good enough to warrant taking a change like this. What is your opinion?

This weekend I will try some improvements to the current TTable instead of throwing it out.

lznncache = use board.hash instead of the feature planes
lznott = same as above but also disable TTable.
```
lzbase-t-4-time-50 v lznncache-t-4-time-50 (161/1000 games)
board size: 19   komi: 7.5
                        wins              black         white       avg cpu
lzbase-t-4-time-50        78 48.45%       44 54.32%     34 42.50%    153.55
lznncache-t-4-time-50     83 51.55%       46 57.50%     37 45.68%    155.84
                                          90 55.90%     71 44.10%

lzbase-t-4-time-50 v lznott-t-4-time-50 (162/1000 games)
board size: 19   komi: 7.5
                     wins              black         white       avg cpu
lzbase-t-4-time-50     75 46.30%       36 44.44%     39 48.15%    141.80
lznott-t-4-time-50     87 53.70%       42 51.85%     45 55.56%    143.82
                                       78 48.15%     84 51.85%
```

**Comment by killerducky** [comment]
Final results. lznott ended up at 57.14%.

```
lzbase-t-4-time-50 v lznncache-t-4-time-50 (426/1000 games)
board size: 19   komi: 7.5
                        wins              black          white        avg cpu
lzbase-t-4-time-50       223 52.35%       114 53.52%     109 51.17%    148.53
lznncache-t-4-time-50    203 47.65%       104 48.83%     99  46.48%    150.72
                                          218 51.17%     208 48.83%

lzbase-t-4-time-50 v lznott-t-4-time-50 (427/1000 games)
board size: 19   komi: 7.5
                     wins              black          white        avg cpu
lzbase-t-4-time-50    183 42.86%       88  41.12%     95  44.60%    146.87
lznott-t-4-time-50    244 57.14%       118 55.40%     126 58.88%    148.92
                                       206 48.24%     221 51.76%
```

**Comment by killerducky** [comment]
It will be weaker when used with --playouts instead of run on equal time. Do you want me to add some command line switches to control these changes? For example `--nncache [strict|relaxed]`, `--ttable [on|off]`, and default them to the current behavior. That would also allow retuning if there is some interaction with the FPU, cfg_puct, and other parameters.

Downside is we might accumulate too many switches. They could be made undocumented and/or controlled by defines, similar to --puct.

**Comment by gcp** [comment]
>It will be weaker when used with --playouts instead of run on equal time

Are you sure? 57% winrate for a CPU usage ratio of 148 vs 146 seems to indicate it may be stronger?

>Do you want me to add some command line switches to control these changes?

No.

**Comment by killerducky** [comment]
That test was done giving each side 0.5s to move, it did not use --playouts. That's why the CPU usage is so similar. I did a test earlier with --playouts and it was faster but worse, I lost the results but I can run again if you like.

**Comment by gcp** [comment]
Wait, so the TTable is worse because of the memory probing overhead, not because of the influence on the search?

Fixed playout with -t1 would be good. I don't mind that it would get a bit weaker at fixed playouts if it's overall more efficient.

**Comment by killerducky** [comment]
The speed difference comes from the NNCache change -- it goes from 0.1% hits to around 12% hits. 

Disabling TTable causes it to recover similar information through NNCache hits instead. These are cheap, only taking CPU time. But they eat into the playout count.

**Comment by gcp** [comment]
Ok, it makes sense now. Instead of making NNCache return the guaranteed correct result, you make it ignore the move history (but still detect simple ko differences), acting as the TTable did for detecting transpositions. Then you don't take the visit/score from the TTable, but reconstruct them by repeatedly reconstructing the tree and hitting the NNCache.

I'm tempted to agree then that this more an indication that using the information from the TTable differently should eventually give equal or better benefits (as it would avoids the Search<>NNCache pingpong)

**Comment by roy7** [comment]
Since NNCache hits are cheap, is it a problem if we don't count a hit as a playout? Using playouts as a way to control amount of GPU power used, than a consistent strength (search tree size). For search tree size strength limit, --visits could be used?

Just as an idea to recapture those 100 playouts that are "wasted" with TTable going away. --playouts then becomes "new playouts not in cache".

**Comment by roy7** [comment]
Since playouts do need to be counted for the actual search to function, maybe on any given move, a NNCache hit increases the playout cap for that move by 1. So we still do 1600 new lookups, but the search will report back more (say 1800 playouts that move if 200 NNCache hits happened).

**Comment by gcp** [comment]
I don't see how changing definitions around (so they are inconsistent and confusing) makes the program better.

**Comment by roy7** [comment]
I'm just trying to illustrate my thought that we don't need to "lose" playouts when dropping TTable if we adjust the search accordingly. Sorry for the terminology distraction.

**Comment by killerducky** [comment]
Ok here is a cleaned up and rebased version.

File: src/NNCache.cpp [hunk 2]
```diff
@@ -55,28 +39,25 @@ bool NNCache::lookup(const Network::NNPlanes& features, Network::Netresult & res
...
     const auto& entry = iter->second;
-    if (entry->features != features) {
-        // Got a hash collision.
-        ++m_collisions;
-        return false;
-    }
 
     // Found it.
     ++m_hits;
+    if (m_hits%100 == 0) {
```
Comment (gcp) [inline src/NNCache.cpp]: Sounds like leftover debugging.

**Comment by killerducky** [comment]
Here are the test results of `-t 1`. It runs faster but we lose a little Elo in the nott case. The effect is small and I only have 190 games each, I'll let this run a little longer. The previous test on equal cpu time had 426 games and showed we more than gain back the Elo.

```
lzbase-p-1600-t-1 v lznncache-p-1600-t-1 (190/1000 games)
board size: 19   komi: 7.5
                       wins              black         white        avg cpu
lzbase-p-1600-t-1        88 46.32%       40 42.11%     48  50.53%    193.44
lznncache-p-1600-t-1    102 53.68%       47 49.47%     55  57.89%    177.66
                                         87 45.79%     103 54.21%

lzbase-p-1600-t-1 v lznott-p-1600-t-1 (190/1000 games)
board size: 19   komi: 7.5
                    wins              black         white       avg cpu
lzbase-p-1600-t-1    102 53.68%       52 54.74%     50 52.63%    193.28
lznott-p-1600-t-1     88 46.32%       45 47.37%     43 45.26%    171.23
                                      97 51.05%     93 48.95%
```

**Comment by marcocalignano** [comment]
@killerducky Just a liitle question outsude the discussion. To test the different binaries do you use the validation program from my repository branch /bin_option ?

**Comment by killerducky** [comment]
No I use gomill ringmaster. It's pretty flexible but it doesn't include the statistical analysis stuff. I just use some website like http://www.quantitativeskills.com/sisa/index.htm for that.

**Comment by marcocalignano** [comment]
If you want to have SPRT just try that version of validation. You can set different binaries and different option for each binaries, still doing the SPRT. 
If someone test it successful I can do a PR for the next branch

**Comment by gcp** [comment]
My results are compatible with yours, pulling.

```
leelaz v leelaz-nott (124/1000 games)
board size: 19   komi: 7.5
              wins              black         white       avg cpu
leelaz          60 48.39%       30 48.39%     30 48.39%    322.47
leelaz-nott     64 51.61%       32 51.61%     32 51.61%    322.00
                                62 50.00%     62 50.00%

leelaz-1k6 v leelaz-nott-1k6 (124/1000 games)
board size: 19   komi: 7.5
                  wins              black         white       avg cpu
leelaz-1k6          64 51.61%       29 46.77%     35 56.45%    223.99
leelaz-nott-1k6     60 48.39%       27 43.55%     33 53.23%    204.90
                                    56 45.16%     68 54.84%
```

**Pull request merged**